src/dune_rules/dep_conf_eval.ml

```diff
@@ -273,10 +273,19 @@ let named ~expander l =
         | Unnamed dep -> add_sandbox_config acc dep
         | Named (_, l) -> List.fold_left l ~init:acc ~f:add_sandbox_config) )
 
-let unnamed ?(sandbox = Sandbox_config.no_special_requirements) ~expander l =
+let unnamed_get_paths ?(sandbox = Sandbox_config.no_special_requirements)
+    ~expander l =
   let expander = prepare_expander expander in
-  ( List.fold_left l ~init:(Action_builder.return ()) ~f:(fun acc x ->
-        let+ () = acc
-        and+ _x = to_action_builder (dep expander x) in
-        ())
+  ( (let+ paths =
+       List.fold_left l ~init:(Action_builder.return []) ~f:(fun acc x ->
+           let+ acc = acc
+           and+ paths = to_action_builder (dep expander x) in
+           paths :: acc)
+     in
+     Path.Set.of_list (List.concat paths))
   , List.fold_left l ~init:sandbox ~f:add_sandbox_config )
+
+let unnamed ?(sandbox = Sandbox_config.no_special_requirements) ~expander l =
+  let expander = prepare_expander expander in
+  let builder, sandbox = unnamed_get_paths ~sandbox ~expander l in
+  (Action_builder.ignore builder, sandbox)
```

Comment:
Although it requires a bit of duplication, I'd prefer to keep the old implementation of unnamed. This new implementation has the potential to really slow down old builds with rules that expand huge globs.